mesosdriver/executor_driver.go

```diff
@@ -83,23 +82,49 @@ func (driver *ExecutorDriver) unacknowledgedUpdates() (result []executor.Call_Up
 
 // eventLoop is the main handler event loop of the driver. Called from Run()
 func (driver *ExecutorDriver) eventLoop(decoder encoding.Decoder,
-	h events.Handler) (err error) {
+	h events.Handler) error {
+
+	var err error
 
 	log.Info("Entering event loop")
-	ctx := context.TODO()
+	ctx := context.Background()
+
+	event := make(chan error)
 
-	for err == nil && !driver.shouldQuit {
-		// housekeeping
-		driver.sendFailedTasks()
+	go func() {
+		for {
+			var e executor.Event
+			if err := decoder.Decode(&e); err == nil {
+				err = h.HandleEvent(ctx, &e)
+			}
 
-		var e executor.Event
-		if err = decoder.Decode(&e); err == nil {
-			err = h.HandleEvent(ctx, &e)
+			select {
+			case event <- err:
+			case <-driver.quitChan:
+				return
+			}
+		}
+	}()
+
+	for {
+		select {
+		case <-driver.quitChan:
+			log.Info("Event loop canceled")
+			return nil
+		case err = <-event:
```

Comment:
`err` is propagated here from the top level scope. The other goroutine has its own `err`.